Assets/_BForBoss/_Character/Scripts/PlayerWallRunBehaviour.cs

```diff
@@ -0,0 +1,232 @@
+        private float _baseMaxSpeed;
+        private Func<Vector2> _movementInput;
+
+        private float _currentJumpDuration = 0f;
+        private float _timeSinceWallAttach = 0f;
+        private float _timeSinceWallDetach = 0f;
+
+        private Transform ChildTransform
+        {
+            get
+            {
+                return _fpsCharacter != null ? _fpsCharacter.rootPivot : transform;
+            }
+        }
+
+        public void Initialize(Character baseCharacter, Func<Vector2> getMovementInput)
+        {
+            _baseCharacter = baseCharacter;
+            _fpsCharacter = baseCharacter as FirstPersonCharacter;
+            _movementInput = getMovementInput;
+            _mask = LayerMask.GetMask("ParkourWall");
+        }
+
+        public void Falling(Vector3 _)
+        {
+            if (!CanWallRun()) return;
+            _hits = new RaycastHit[directions.Length];
+            for (int i = 0; i < directions.Length; i++)
+            {
+                Vector3 dir = ChildTransform.TransformDirection(directions[i]); //convert directions to relative to player
+                Physics.Raycast(ChildTransform.position, dir, out _hits[i], _wallMaxDistance, _mask);
+                if (_hits[i].collider != null) Debug.DrawRay(ChildTransform.position, dir * _hits[i].distance, Color.green);
+                else Debug.DrawRay(ChildTransform.position, dir * _wallMaxDistance, Color.red);
+            }
+
+            if (GetSmallestRaycastHit(_hits, out RaycastHit hit))
+            {
+                WallRun(hit);
+            }
+            else if (_isWallRunning) StopWallRunning();
+        }
+
+        public void OnJumped(ref int _jumpCount)
+        {
+            _currentJumpDuration = 0f;
+            if (_isWallRunning)
+            {
+                StopWallRunning();
+                _jumpCount = 0;
+            }
+        }
+
+        public void OnLanded()
+        {
+            _currentJumpDuration = 0f;
+            if (_isWallRunning)
+            {
+                StopWallRunning();
+            }
+        }
+
+        public bool CanJump()
+        {
+            return _isWallRunning; // can always jump out of a wall run
+        }
+
+        public void OnWallRunning()
+        {
+            if (!_isWallRunning)
+            {
+                _timeSinceWallDetach += Time.deltaTime;
+                return;
+            }
+
+            if (_movementInput().sqrMagnitude <= 0)
+            {
+                return;
+            }
+            _timeSinceWallAttach += Time.deltaTime;
+            var velocity = _baseCharacter.GetVelocity();
+            var alongWall = ChildTransform.TransformDirection(Vector3.forward).normalized;
+            velocity = velocity.dot(alongWall) * alongWall;
+            velocity += Vector3.down * _wallGravityDownForce * Time.deltaTime;
+            _baseCharacter.SetVelocity(velocity);
+        }
+
+        public bool CalcJumpVelocity(out Vector3 velocity)
```

Comment:
Why do you need an `out ` here? why not just create a new Vector3 and return that, since it starts at Vector3.zero anyway?

Reply:
Its how I was overriding things in the first person controller, since I only want to use the wallrunbehaviour's jump velocity if we're wall running. Using an out saves a vector comparison

Reply:
yeah alright, i think it keeps it consistent